File: backend/src/test/java/com/coing/domain/coin/market/service/MarketServiceTest.java
package com.coing.domain.coin.market.service;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import com.coing.domain.coin.market.dto.MarketDto;
import com.coing.domain.coin.market.dto.MarketResponseDto;
import com.coing.domain.coin.market.entity.Market;
import com.coing.domain.coin.market.repository.MarketRepository;

@ExtendWith(MockitoExtension.class)
public class MarketServiceTest {

	@Mock
	private MarketRepository marketRepository;

	@Mock
	private RestTemplate restTemplate;

	@Mock
	private MarketCacheService marketCacheService;

	@InjectMocks
	private MarketService marketService;

	@Value("${upbit.market.uri}")
	private String UPBIT_MARKET_URI;

	private MarketDto[] mockMarketDtos;

	@BeforeEach
	void setUp() {
		// Test setup: initialize mock data
		mockMarketDtos = new MarketDto[] {
			new MarketDto("KRW-BTC", "비트코인", "BTC"),
			new MarketDto("KRW-ETH", "이더리움", "ETH"),
		};
	}

	@Test
	@DisplayName("t1: 코인 목록 자동 갱신 - 정상 동작 테스트")
	void testUpdateMarketList_Success() {
		// Given
		ResponseEntity<MarketDto[]> responseEntity = ResponseEntity.ok(mockMarketDtos);
		when(restTemplate.getForEntity(UPBIT_MARKET_URI, MarketDto[].class)).thenReturn(responseEntity);

		// When
		marketService.updateMarketList();

		// Then
		verify(marketRepository, times(1)).saveAll(anyList());
	}

	@Test
	@DisplayName("t2: 코인 목록 자동 갱신 - 외부 API 에러 시 DB fallback 테스트")
	void testUpdateMarketList_Exception() {
		// Given
		when(restTemplate.getForEntity(UPBIT_MARKET_URI, MarketDto[].class)).thenThrow(
			new RuntimeException("Rest API Error"));

		List<Market> dbMarkets = Arrays.asList(
			new Market("KRW-BTC", "비트코인", "BTC"),
			new Market("KRW-ETH", "이더리움", "ETH")
		);
		when(marketRepository.findAll()).thenReturn(dbMarkets);

		// When & Then
		marketService.updateMarketList();

		verify(marketRepository, times(0)).saveAll(anyList());
		verify(marketCacheService, times(1)).updateMarketCache(dbMarkets);
	}

	@Test
	@DisplayName("t3: 코인 목록 갱신 요청 - 정상 동작 테스트")
	void testRefreshMarketList_Success() {
		// Given
		ResponseEntity<MarketDto[]> responseEntity = ResponseEntity.ok(mockMarketDtos);
		when(restTemplate.getForEntity(UPBIT_MARKET_URI, MarketDto[].class)).thenReturn(responseEntity);

		// When
		marketService.refreshMarketList();

		// Then
		verify(marketRepository, times(1)).saveAll(anyList());
	}

	@Test
	@DisplayName("t4: 코인 목록 전체 조회 - 정상 동작 테스트")
	void testGetMarkets() {
		// Given
		Map<String, Market> mockMarketMap = Map.of(
			"KRW-BTC", new Market("KRW-BTC", "비트코인", "Bitcoin"),
			"KRW-ETH", new Market("KRW-ETH", "이더리움", "Ethereum")
		);

		when(marketCacheService.getCachedMarketMap()).thenReturn(mockMarketMap);
		Pageable pageable = PageRequest.of(0, 10);

		// When
		Page<Market> markets = marketService.getMarkets(pageable);

		// Then
		assertThat(markets.getContent()).hasSize(2);
		assertThat(markets.getTotalElements()).isEqualTo(2);
	}

	@Test
	@DisplayName("t5: 기준 통화별 마켓 목록 조회 - 정상 동작 테스트")
	void getAllCoinsByMarket_ShouldReturnFilteredMarkets() {
		// given
		String type = "KRW";
		Map<String, Market> mockMarketMap = Map.of(
			"KRW-BTC", new Market("KRW-BTC", "비트코인", "Bitcoin"),
			"KRW-ETH", new Market("KRW-ETH", "이더리움", "Ethereum")
		);

		when(marketCacheService.getCachedMarketMap()).thenReturn(mockMarketMap);
		Pageable pageable = PageRequest.of(0, 10);

		// when
		Page<MarketResponseDto> result = marketService.getAllMarketsByQuote(null, type, pageable);

		// then
		assertThat(result.getContent()).hasSize(2);
		assertThat(result.getContent().get(0).code()).isEqualTo("KRW-BTC");
		assertThat(result.getContent().get(1).code()).isEqualTo("KRW-ETH");
	}

	@Test
	@DisplayName("t5: 기준 통화별 마켓 목록 조회 - 정확한 기준 통화명이 아닐 시 빈 리스트 반환")
	void getAllCoinsByMarket_ShouldReturnEmptyList_WhenNoMatchingMarkets() {
		// given
		String type = "USD";
		Pageable pageable = PageRequest.of(0, 10);
		when(marketCacheService.getCachedMarketMap()).thenReturn(Collections.emptyMap());

		// when
		Page<MarketResponseDto> result = marketService.getAllMarketsByQuote(null, type, pageable);

		// then
		assertThat(result.getContent()).isEmpty();
	}
}

File: backend/src/test/java/com/coing/domain/coin/market/service/MarketServiceTest.kt
package com.coing.domain.coin.market.service

import com.coing.domain.bookmark.repository.BookmarkRepository
import com.coing.domain.coin.market.dto.MarketResponseDto
import com.coing.domain.coin.market.entity.Market
import com.coing.domain.coin.market.port.MarketDataPort
import com.coing.domain.coin.market.repository.MarketRepository
import com.coing.domain.user.dto.CustomUserPrincipal
import com.coing.global.exception.BusinessException
import com.coing.util.MessageUtil
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import java.util.*

class MarketServiceTest {

	private lateinit var marketService: MarketService
	private lateinit var messageUtil: MessageUtil
	private lateinit var marketCacheService: MarketCacheService
	private lateinit var bookmarkRepository: BookmarkRepository
	private lateinit var marketRepository: MarketRepository
	private lateinit var marketDataPort: MarketDataPort

	private val testMarket = Market(
		code = "KRW-BTC",
		koreanName = "비트코인",
		englishName = "Bitcoin"
	)

	@BeforeEach
	fun setUp() {
		messageUtil = mock(MessageUtil::class.java)
		marketCacheService = mock(MarketCacheService::class.java)
		bookmarkRepository = mock(BookmarkRepository::class.java)
		marketRepository = mock(MarketRepository::class.java)
		marketDataPort = mock(MarketDataPort::class.java)

		marketService = MarketService(
			messageUtil,
			marketCacheService,
			bookmarkRepository,
			marketRepository,
			marketDataPort
		)
	}

	@Test
	fun `getCachedMarketByCode - ok`() {
		val cache = mapOf(testMarket.code to testMarket)
		`when`(marketCacheService.getCachedMarketMap()).thenReturn(cache)

		val result = marketService.getCachedMarketByCode("KRW-BTC")
		assertThat(result).isEqualTo(testMarket)
	}

	@Test
	fun `getCachedMarketByCode - exception when market not found`() {
		`when`(marketCacheService.getCachedMarketMap()).thenReturn(emptyMap())
		`when`(messageUtil.resolveMessage("market.not.found")).thenReturn("Market not found")

		val exception = catchThrowable {
			marketService.getCachedMarketByCode("KRW-BTC")
		}

		assertThat(exception).isInstanceOf(BusinessException::class.java)
		assertThat((exception as BusinessException).status).isEqualTo(HttpStatus.NOT_FOUND)
	}

	@Test
	fun `getMarketByUserAndCode - ok`() {
		val principal = mock(CustomUserPrincipal::class.java)
		val id = UUID.randomUUID()
		`when`(principal.id).thenReturn(id)
		`when`(bookmarkRepository.existsByUserIdAndMarketCode(id, "KRW-BTC")).thenReturn(true)
		`when`(marketCacheService.getCachedMarketMap()).thenReturn(mapOf("KRW-BTC" to testMarket))

		val result = marketService.getMarketByUserAndCode(principal, "KRW-BTC")
		assertThat(result).isInstanceOf(MarketResponseDto::class.java)
		assertThat(result.code).isEqualTo("KRW-BTC")
		assertThat(result.isBookmarked).isTrue()
	}

	@Test
	fun `getAllMarketsByQuote - filtering with bookmark`() {
		// given
		val principal = mock(CustomUserPrincipal::class.java)
		val id = UUID.randomUUID()
		`when`(principal.id).thenReturn(id)

		// 마켓 객체 mock
		val market = mock(Market::class.java)
		`when`(market.code).thenReturn("KRW-BTC")
		`when`(market.koreanName).thenReturn("비트코인")
		`when`(market.englishName).thenReturn("BTC")

		// 캐시에서 마켓 맵 반환
		`when`(marketCacheService.getCachedMarketMap()).thenReturn(
			mapOf("KRW-BTC" to market)
		)

		// 북마크 객체 mock
		val bookmark = mock(com.coing.domain.bookmark.entity.Bookmark::class.java)
		`when`(bookmark.market).thenReturn(market)

		`when`(bookmarkRepository.findByUserIdAndQuote(id, "KRW"))
			.thenReturn(listOf(bookmark))

		val pageable: Pageable = PageRequest.of(0, 10)

		// when
		val result = marketService.getAllMarketsByQuote(principal, "KRW", pageable)

		// then
		assertThat(result.content).hasSize(1)
		assertThat(result.content.first().isBookmarked).isTrue()
	}
}
